packages/solid-query/src/createBaseQuery.ts

```diff
@@ -35,46 +42,99 @@ export function createBaseQuery<
   const queryClient = createMemo(() =>
     useQueryClient({ context: options().context }),
   )
-  const emptyData = Symbol('empty')
 
   const defaultedOptions = queryClient().defaultQueryOptions(options())
   defaultedOptions._optimisticResults = 'optimistic'
   const observer = new Observer(queryClient(), defaultedOptions)
 
+  const emptyData = Symbol('empty')
+
+  let resolver: (value: TData | undefined) => void
+  let queryResolver: (value: DehydratedState) => void
+
+  // This resource will be used in a Server environment to
+  // dehydrate the queryClient when the query is
+  // pre fetched on the server. It will always be undefined
+  // if an observer is mounted on client.
+  const [queryResource] = createResource<DehydratedState | undefined>(() => {
+    return new Promise((resolve) => {
+      if (!isServer) {
+        resolve(undefined)
+      }
+      queryResolver = resolve
+    })
+  })
+
+  // If queryResource is defined,
+  // This means that the query was fetched on the server!
+  // We need to hydrate the queryClient with the query from
+  // the server before we set up the observer and its results.
+  if (queryResource()) {
+    hydrate(queryClient(), queryResource())
+  }
+
   const [state, setState] = createStore<QueryObserverResult<TData, TError>>(
     observer.getOptimisticResult(defaultedOptions),
   )
 
-  const [dataResource, { refetch, mutate }] = createResource<TData | undefined>(
+  const [dataResource, { mutate, refetch }] = createResource<TData | undefined>(
     () => {
       return new Promise((resolve) => {
-        if (!(state.isFetching && state.isLoading)) {
-          if ((unwrap(state.data) as TData | typeof emptyData) === emptyData) {
-            resolve(undefined)
+        if (isServer) {
+          if (!(state.isFetching && state.isLoading)) {
```

Comment:
`isFetching && isLoading` could be unified to just check for `isInitialLoading`

Reply:
Yeah since staleTime is zero when the query is mounted on the client it will do a refetch. You could opt out of it by having a non zero staletime. Or maybe just setting refetchOnMount to false as well. I'll update the docs with this caveat and also change the checks to `isInitialLoading`

Reply:
non-zero staleTime is the best way imo 👍

Reply:
I'll look more into this! I'm not fully sure how it works but maybe using `isRestoring` for the first render on the client might help? Just spitballing here. I'll test more and report back

Reply:
`useIsRestoring` is currently only used for hydrating from the persist plugin, not when hydrating via the `Hydration` boundary.

Reply:
Okay I think I got it working so that a query is not refetched when a query is mounted from the server. This change [ad45d25](https://github.com/TanStack/query/pull/4803/commits/ad45d259812fd98baed974612206df8e7703bb3d)
makes solid-query work exactly the same between CSR and SSR'd apps!